pkg/reconciler/pipelinerun/resources/pipelinerunresolution.go

```diff
@@ -422,6 +479,45 @@ func (t *ResolvedPipelineTask) skipBecauseResultReferencesAreMissing(facts *Pipe
 	return false
 }
 
+// skipBecausePipelineRunPipelineTimeoutReached returns true if the task shouldn't be launched because the elapsed time since
+// the PipelineRun started is greater than the PipelineRun's pipeline timeout
+func (t *ResolvedPipelineTask) skipBecausePipelineRunPipelineTimeoutReached(facts *PipelineRunFacts) bool {
+	if t.checkParentsDone(facts) {
+		if facts.PipelineTimeout != nil && *facts.PipelineTimeout != config.NoTimeoutDuration && facts.StartTime != nil {
+			// If the elapsed time since the PipelineRun's start time is greater than the the PipelineRun's Pipeline timeout, skip.
+			return facts.Clock.Since(*facts.StartTime) > *facts.PipelineTimeout
+		}
+	}
+
+	return false
+}
+
+// skipBecausePipelineRunTasksTimeoutReached returns true if the task shouldn't be launched because the elapsed time since
+// the PipelineRun started is greater than the PipelineRun's tasks timeout
+func (t *ResolvedPipelineTask) skipBecausePipelineRunTasksTimeoutReached(facts *PipelineRunFacts) bool {
+	if t.checkParentsDone(facts) && !t.IsFinalTask(facts) {
+		if facts.TasksTimeout != nil && *facts.TasksTimeout != config.NoTimeoutDuration && facts.StartTime != nil {
+			// If the elapsed time since the PipelineRun's start time is greater than the the PipelineRun's Tasks timeout, skip.
+			return facts.Clock.Since(*facts.StartTime) > *facts.TasksTimeout
+		}
+	}
+
+	return false
+}
+
+// skipBecausePipelineRunFinallyTimeoutReached returns true if the task shouldn't be launched because the elapsed time since
+// finally tasks started being executed is greater than the PipelineRun's finally timeout
+func (t *ResolvedPipelineTask) skipBecausePipelineRunFinallyTimeoutReached(facts *PipelineRunFacts) bool {
+	if t.checkParentsDone(facts) && t.IsFinalTask(facts) {
+		if facts.FinallyTimeout != nil && *facts.FinallyTimeout != config.NoTimeoutDuration && facts.FinallyStartTime != nil {
+			// If the elapsed time since the PipelineRun's finally start time is greater than the the PipelineRun's finally timeout, skip.
+			return facts.Clock.Since(*facts.FinallyStartTime) > *facts.FinallyTimeout
+		}
+	}
+
+	return false
+}
+
```

Comment:
these scenarios for not scheduling a taskrun or run because of timeouts could be included in "stopping mode" - this will also prevent complicating the skipping logic further (already handles "stopping mode") - what do you think @abayer?

https://github.com/abayer/tektoncd-pipeline/blob/8b740d50ef106c9cf633634851615ac316289a54/pkg/reconciler/pipelinerun/resources/pipelinerunstate.go#L357-L368

Reply:
Good question! I _think_ I prefer having the explicit checks for `.timeouts.tasks`, `.timeouts.finally`, and `.timeouts.pipeline` - that lets us set a more specific reason for skipping pertinent tasks, but I'm willing to be persuaded otherwise.